File: WallpaperRotator.cs
// WallpaperRotator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Diagnostics;

class WallpaperRotator
{
    private string dir;
    private List<string> images = new List<string>();
    private int interval = 60;
    private bool randomOrder = false;
    private int currentIndex = 0;
    private bool running = false;
    private string currentWall = null;
    private Timer timer;
    private readonly string stateFile = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".wallpaper_rotator_state.txt");

    public WallpaperRotator() => LoadState();

    private void LoadState()
    {
        if (!File.Exists(stateFile)) return;
        var lines = File.ReadAllLines(stateFile);
        if (lines.Length >= 4)
        {
            if (!string.IsNullOrEmpty(lines[0])) dir = lines[0];
            int.TryParse(lines[1], out interval);
            bool.TryParse(lines[2], out randomOrder);
            if (!string.IsNullOrEmpty(lines[3])) currentWall = lines[3];
            ScanImages();
        }
    }

    private void SaveState()
    {
        File.WriteAllLines(stateFile, new string[] {
            dir ?? "",
            interval.ToString(),
            randomOrder.ToString(),
            currentWall ?? ""
        });
    }

    private void ScanImages()
    {
        if (string.IsNullOrEmpty(dir) || !Directory.Exists(dir))
        {
            images.Clear();
            return;
        }
        var exts = new HashSet<string> { ".jpg", ".jpeg", ".png", ".bmp", ".gif", ".tiff" };
        images = Directory.GetFiles(dir)
            .Where(f => exts.Contains(Path.GetExtension(f).ToLower()))
            .ToList();
    }

    private bool SetWallpaper(string path)
    {
        if (!File.Exists(path))
        {
            Console.WriteLine($"File not found: {path}");
            return false;
        }
        try
        {
            var os = Environment.OSVersion.Platform;
            if (os == PlatformID.Win32NT)
            {
                Process.Start("reg", $"add \"HKEY_CURRENT_USER\\Control Panel\\Desktop\" /v Wallpaper /t REG_SZ /d \"{path}\" /f");
                Process.Start("RUNDLL32.EXE", "user32.dll,UpdatePerUserSystemParameters 1 True");
            }
            else if (os == PlatformID.MacOSX)
            {
                Process.Start("osascript", $"-e 'tell application \"Finder\" to set desktop picture to POSIX file \"{path}\"'");
            }
            else // Linux
            {
                // Try GNOME
                var p = Process.Start("gsettings", "get org.gnome.desktop.background picture-uri");
                p.WaitForExit();
                if (p.ExitCode == 0)
                    Process.Start("gsettings", $"set org.gnome.desktop.background picture-uri \"file://{path}\"");
                else
                    Process.Start("feh", $"--bg-scale \"{path}\"");
            }
            currentWall = path;
            SaveState();
            return true;
        }
        catch (Exception e)
        {
            Console.WriteLine($"Error setting wallpaper: {e.Message}");
            return false;
        }
    }

    public string GetCurrentWallpaper()
    {
        if (!string.IsNullOrEmpty(currentWall)) return currentWall;
        var os = Environment.OSVersion.Platform;
        try
        {
            if (os == PlatformID.MacOSX)
            {
                var p = Process.Start("osascript", "-e 'tell application \"Finder\" to get desktop picture'");
                p.WaitForExit();
                return p.StandardOutput.ReadToEnd().Trim();
            }
            else if (os == PlatformID.Unix) // Linux
            {
                var p = Process.Start("gsettings", "get org.gnome.desktop.background picture-uri");
                p.WaitForExit();
                string line = p.StandardOutput.ReadToEnd().Trim().Trim('\'');
                if (line.StartsWith("file://")) return line.Substring(7);
            }
        }
        catch { }
        return "Unknown";
    }

    public void NextWallpaper()
    {
        if (images.Count == 0)
        {
            Console.WriteLine("No images in directory. Please add images.");
            return;
        }
        int idx = randomOrder ? new Random().Next(images.Count) :
            (currentIndex = (currentIndex + 1) % images.Count);
        string path = images[idx];
        if (SetWallpaper(path))
            Console.WriteLine($"Wallpaper changed to: {path}");
    }

    private void TimerCallback(object state) => NextWallpaper();

    public void Start()
    {
        if (running)
        {
            Console.WriteLine("Already running.");
            return;
        }
        if (string.IsNullOrEmpty(dir) || !Directory.Exists(dir))
        {
            Console.WriteLine("Please set a directory with --add-dir");
            return;
        }
        ScanImages();
        if (images.Count == 0)
        {
            Console.WriteLine("No images found in directory.");
            return;
        }
        running = true;
        timer = new Timer(TimerCallback, null, TimeSpan.FromSeconds(interval), TimeSpan.FromSeconds(interval));
        Console.WriteLine($"Started rotation every {interval} seconds.");
        SaveState();
    }

    public void Stop()
    {
        timer?.Dispose();
        timer = null;
        running = false;
        Console.WriteLine("Stopped rotation.");
    }

    public void SetDirectory(string path)
    {
        if (!Directory.Exists(path))
        {
            Console.WriteLine($"Directory not found: {path}");
            return;
        }
        dir = path;
        ScanImages();
        currentIndex = 0;
        SaveState();
        Console.WriteLine($"Directory set to: {path} ({images.Count} images)");
    }

    static void Main(string[] args)
    {
        var opts = new Dictionary<string, string>();
        for (int i = 0; i < args.Length; i++)
        {
            if (args[i].StartsWith("--"))
            {
                string key = args[i].Substring(2);
                if (i + 1 < args.Length && !args[i + 1].StartsWith("--"))
                    opts[key] = args[++i];
                else
                    opts[key] = "";
            }
        }

        var rotator = new WallpaperRotator();
        if (opts.ContainsKey("add-dir")) rotator.SetDirectory(opts["add-dir"]);
        if (opts.ContainsKey("interval"))
        {
            rotator.interval = int.Parse(opts["interval"]);
            rotator.SaveState();
        }
        if (opts.ContainsKey("random"))
        {
            rotator.randomOrder = true;
            rotator.SaveState();
        }

        if (opts.ContainsKey("start"))
        {
            rotator.Start();
            // keep alive
            while (rotator.running)
                Thread.Sleep(1000);
        }
        else if (opts.ContainsKey("stop"))
            rotator.Stop();
        else if (opts.ContainsKey("next"))
            rotator.NextWallpaper();
        else if (opts.ContainsKey("current"))
            Console.WriteLine($"Current wallpaper: {rotator.GetCurrentWallpaper()}");
        else
            Console.WriteLine("No command given. Use --help.");
    }
}

File: WallpaperRotator.java
// WallpaperRotator.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

public class WallpaperRotator {
    private String dir;
    private List<String> images = new ArrayList<>();
    private int interval = 60;
    private boolean randomOrder = false;
    private int currentIndex = 0;
    private boolean running = false;
    private String currentWall = null;
    private ScheduledExecutorService scheduler;
    private final String stateFile = System.getProperty("user.home") + "/.wallpaper_rotator_state.txt";

    public WallpaperRotator() {
        loadState();
    }

    private void loadState() {
        try {
            Path path = Paths.get(stateFile);
            if (Files.exists(path)) {
                List<String> lines = Files.readAllLines(path);
                if (lines.size() >= 4) {
                    if (!lines.get(0).isEmpty()) dir = lines.get(0);
                    interval = Integer.parseInt(lines.get(1));
                    randomOrder = Boolean.parseBoolean(lines.get(2));
                    if (!lines.get(3).isEmpty()) currentWall = lines.get(3);
                    scanImages();
                }
            }
        } catch (Exception e) {}
    }

    private void saveState() {
        try {
            String content = String.join("\n", 
                dir == null ? "" : dir,
                String.valueOf(interval),
                String.valueOf(randomOrder),
                currentWall == null ? "" : currentWall
            );
            Files.write(Paths.get(stateFile), content.getBytes());
        } catch (Exception e) {}
    }

    private void scanImages() {
        if (dir == null || !Files.isDirectory(Paths.get(dir))) {
            images.clear();
            return;
        }
        String[] exts = {".jpg", ".jpeg", ".png", ".bmp", ".gif", ".tiff"};
        try {
            images = Files.list(Paths.get(dir))
                .filter(Files::isRegularFile)
                .map(Path::toString)
                .filter(p -> {
                    String ext = p.substring(p.lastIndexOf('.')).toLowerCase();
                    return Arrays.asList(exts).contains(ext);
                })
                .collect(Collectors.toList());
        } catch (IOException e) {
            images.clear();
        }
    }

    private boolean setWallpaper(String path) {
        if (!Files.exists(Paths.get(path))) {
            System.out.println("File not found: " + path);
            return false;
        }
        String os = System.getProperty("os.name").toLowerCase();
        try {
            if (os.contains("win")) {
                Runtime.getRuntime().exec("reg add \"HKEY_CURRENT_USER\\Control Panel\\Desktop\" /v Wallpaper /t REG_SZ /d \"" + path + "\" /f");
                Runtime.getRuntime().exec("RUNDLL32.EXE user32.dll,UpdatePerUserSystemParameters 1 True");
            } else if (os.contains("mac")) {
                String script = "tell application \"Finder\" to set desktop picture to POSIX file \"" + path + "\"";
                Runtime.getRuntime().exec(new String[]{"osascript", "-e", script});
            } else {
                // Linux
                Process p = Runtime.getRuntime().exec(new String[]{"gsettings", "get", "org.gnome.desktop.background", "picture-uri"});
                if (p.waitFor() == 0) {
                    Runtime.getRuntime().exec(new String[]{"gsettings", "set", "org.gnome.desktop.background", "picture-uri", "file://" + path});
                } else {
                    Runtime.getRuntime().exec(new String[]{"feh", "--bg-scale", path});
                }
            }
            currentWall = path;
            saveState();
            return true;
        } catch (Exception e) {
            System.err.println("Error setting wallpaper: " + e.getMessage());
            return false;
        }
    }

    public String getCurrentWallpaper() {
        if (currentWall != null) return currentWall;
        String os = System.getProperty("os.name").toLowerCase();
        try {
            if (os.contains("mac")) {
                Process p = Runtime.getRuntime().exec(new String[]{"osascript", "-e", "tell application \"Finder\" to get desktop picture"});
                try (BufferedReader br = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
                    return br.readLine().trim();
                }
            } else if (os.contains("linux")) {
                Process p = Runtime.getRuntime().exec(new String[]{"gsettings", "get", "org.gnome.desktop.background", "picture-uri"});
                try (BufferedReader br = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
                    String line = br.readLine();
                    if (line != null) {
                        String uri = line.trim().replace("'", "");
                        if (uri.startsWith("file://")) return uri.substring(7);
                    }
                }
            }
        } catch (Exception e) {}
        return "Unknown";
    }

    public void nextWallpaper() {
        if (images.isEmpty()) {
            System.out.println("No images in directory. Please add images.");
            return;
        }
        int idx = randomOrder ? new Random().nextInt(images.size()) :
            (currentIndex = (currentIndex + 1) % images.size());
        String path = images.get(idx);
        if (setWallpaper(path)) {
            System.out.println("Wallpaper changed to: " + path);
        }
    }

    private void rotationLoop() {
        scheduler = Executors.newSingleThreadScheduledExecutor();
        scheduler.scheduleAtFixedRate(this::nextWallpaper, interval, interval, TimeUnit.SECONDS);
    }

    public void start() {
        if (running) {
            System.out.println("Already running.");
            return;
        }
        if (dir == null || !Files.isDirectory(Paths.get(dir))) {
            System.out.println("Please set a directory with --add-dir");
            return;
        }
        scanImages();
        if (images.isEmpty()) {
            System.out.println("No images found in directory.");
            return;
        }
        running = true;
        rotationLoop();
        System.out.println("Started rotation every " + interval + " seconds.");
        saveState();
    }

    public void stop() {
        if (scheduler != null) {
            scheduler.shutdownNow();
            scheduler = null;
        }
        running = false;
        System.out.println("Stopped rotation.");
    }

    public void setDirectory(String path) {
        if (!Files.isDirectory(Paths.get(path))) {
            System.out.println("Directory not found: " + path);
            return;
        }
        dir = path;
        scanImages();
        currentIndex = 0;
        saveState();
        System.out.println("Directory set to: " + path + " (" + images.size() + " images)");
    }

    public static void main(String[] args) throws Exception {
        Map<String, String> opts = new HashMap<>();
        for (int i = 0; i < args.length; i++) {
            if (args[i].startsWith("--")) {
                String key = args[i].substring(2);
                if (i+1 < args.length && !args[i+1].startsWith("--")) {
                    opts.put(key, args[++i]);
                } else {
                    opts.put(key, "");
                }
            }
        }
        WallpaperRotator rotator = new WallpaperRotator();
        if (opts.containsKey("add-dir")) rotator.setDirectory(opts.get("add-dir"));
        if (opts.containsKey("interval")) {
            rotator.interval = Integer.parseInt(opts.get("interval"));
            rotator.saveState();
        }
        if (opts.containsKey("random")) {
            rotator.randomOrder = true;
            rotator.saveState();
        }
        if (opts.containsKey("start")) {
            rotator.start();
            // keep alive
            while (rotator.running) {
                Thread.sleep(1000);
            }
        } else if (opts.containsKey("stop")) {
            rotator.stop();
        } else if (opts.containsKey("next")) {
            rotator.nextWallpaper();
        } else if (opts.containsKey("current")) {
            System.out.println("Current wallpaper: " + rotator.getCurrentWallpaper());
        } else {
            System.out.println("No command given. Use --help.");
        }
    }
}
